### ISSUE TITLE
Climate never shows correct setpoint when set by the schedule (or something similar)

### ISSUE THREAD

**Issue by alexruffell** [home-assistant/frontend#17880]
### Checklist

- [X] I have updated to the latest available Home Assistant version.
- [X] I have cleared the cache of my browser.
- [X] I have tried a different browser to see if it is related to my browser.

### Describe the issue you are experiencing

This has been an ongoing issue for years. I thought it would get fixed with the awesome work done on the climate card but the issue persists.

Background: I am in the USA but have my system set to Metric and English GB as I am European. I have 2 ecoBee 3 thermostats set to metric as well. The setpoints can be increased/decreased in 0.5C steps

<img width="438" alt="image" src="https://github.com/home-assistant/frontend/assets/4006944/5b3505d8-f61d-4824-84b1-4c6adea97ab6"><img width="439" alt="image" src="https://github.com/home-assistant/frontend/assets/4006944/69d152f3-55ab-428e-a0f3-c2c09a639a8c">

<img width="445" alt="image" src="https://github.com/home-assistant/frontend/assets/4006944/42c5f8b0-a524-4743-b039-9650be091e65"><img width="444" alt="image" src="https://github.com/home-assistant/frontend/assets/4006944/63bf7de8-07c8-441c-8e1b-65bd13b13e68">

Also, if you move the slider the setpoint clicks back to 0.5C steps, but if you use the buttons + & - then it moves up and down by adding/subtracting 0.5C therefore the setpoint retains whatever offset it had.

Starting point 23.6C:
Slider UP: 24... 24.5
Button UP: 24.1... 24.6 (if you use the slider first, then the button works well as the starting point is already correct).

I've been told before it was due to some conversion... but that makes absolutely no sense for a "setpoint" as it is a static value set from the UI and should match across interfaces. I know ecoBee rounds up/down the readings showing temperatures only in 0.5C increments, while HA supposedly shows the actual value - that is great - but this argument cannot be applied to setpoints.

Also, it appears that the incorrect setpoint happens when it is set by the schedule because when I change it from the thermostat, HA shows it correctly at 0.5C intervals.

### Describe the behavior you expected

At the very least, both the slider and buttons should snap the setpoint to a 0.5V step (slider does, buttons don't).

If the incorrect setpoint is due to ecoBee, please let us (ecoBee users) know so we can ask ecoBee to fix it. Regardless, any way to "snap" it to the proper step? I guess knowing whether to go up or down might be an issue is simple rounding rules don't work.

In short, the expected behavior is that the HA frontend always matches the thermostat and ecoBee app with regards to the setpoint value.

### Steps to reproduce the issue

It happens daily, but I never replicated it manually so this is somewhat of an educated guess
1. Set schedule to 23.5C
2. When schedule is applied, the setpoint in HA should be off and probably at 23.6C (unexpected)
3. Press up/down button and notice that the offset (delta from 0.5V step) is retained (unexpected)
4. Move slider and notice that the offset is eliminated (this is ok)
...

### What version of Home Assistant Core has the issue?

core-2023.9.1

### Additional information

With my luck... this issue is in the wrong repository. At times it is so confusing where stuff is supposed to go!

**Comment by karwosts** [comment]
So if I understand correctly:

1) You set the ecobee setpoint to 23.5
2) It shows up in HA as 23.6? Does the climate attributes also read 23.6 in developer-tools/state?

I'm not exactly sure here but I think you might want a core issue for this, if the setpoint is wrong in the entity attributes than that's a core responsibility to fix. 

Frontend could maybe always force the +/- buttons to snap to half-C increments, but that seems like it only fix half the problem, and is just a bandaid over what seems like the real problem to me. 

You might try enabling your debug logs for climate and ecobee and seeing what it prints out around the time when the setpoint changes to see if you can get any details of what conversion is happening.

**Comment by alexruffell** [comment]
@karwosts Earlier this morning I set both thermostats to 25C using the ecobee app and then left home for a couple hours. When I got back both TS on HA were showing a setpoint of 24.8C even though I am quite sure they were showing 25C in HA when I changed the setpoint via the app. So, the schedule may have nothing to do with the issue as the schedule did not change during the couple hours I was out. I do not have any automations changing the setpoint either.

I now set the setpoint to 24.5C via the app on both thermostats and HA shows 24.5C as expected. I'll now wait to see when it changes and what might be the cause. All I know is that when I look at the setpoint in HA it is most often incorrect.

This is what dev tools shows but right now the setpoint is correct in HA so I will check again next time it is wrong:
<img width="787" alt="image" src="https://github.com/home-assistant/frontend/assets/4006944/29beda49-d3d8-4642-a0e9-327b0c21762f">

**Comment by alexruffell** [comment]
@karwosts - This morning I set the temperatures to a higher temperature using HA and then found that the temperature went back down because the @ home schedule kicked in. Not sure why one was at 23.5 and the other 23.6... maybe someone adjusted the other one. Anyhow Dev Tools is showing 23.6 so I believe you said that means it is a Core issue, right?

<img width="791" alt="image" src="https://github.com/home-assistant/frontend/assets/4006944/c808b75e-c177-40e1-8055-cea913809efb">

**Comment by karwosts** [comment]
I think it would be reasonable to open a core issue for this integration and see if they can figure out why it's getting the conversions incorrect.

**Comment by Woodcat64** [comment]
I tried already. https://github.com/home-assistant/core/issues/100256

The issue is , ( when using `°C `)that the actual nest thermostat, the nest app and nest web UI is rounding to nearest half, but HA is rounding to first decimal. So when the google_nest_sdm sends the set point temperature to HA, it looks looks this `{'coolCelsius': 25.83498}` which HA rounds to `25.8`, while Nest rounds to `26`

This happens when changing setpoint manually on the thermostat, but also from nest schedule and coming out off away/eco  mode.

**Comment by alexruffell** [comment]
@Woodcat64 - While what you describe makes sense for actual temperature as HA would show the most accurate while the thermostat opts to show the closest 0.5C step, it really doesn't make any sense at all for a setpoint. A setpoint is a fixed value so if should not be converted - ever. I set my setpoint in C and I want HA to pull it in C so why would there be any conversions? The only thing I can think of is that I am a European living in the USA so in this land of Imperial/Standard, I go with Metric... possibly breaking something that a programmer did not think about. Can't explain it otherwise...

**Comment by Woodcat64** [comment]
Are you suggesting that, when my Nest thermostat and the app shows 25C setpoint, it is fine that HA decides to show that setpoint is 24.7C of 25.2C? All I'm asking is for HA showing same info as Nest. Also European living in Canada here.

**Comment by alexruffell** [comment]
@Woodcat64 - No I am saying that the setpoint (the temperature you want) should be identical on Nest and HA as it is just a 'target' setting. Instead, the current temperature might be rounded to the nearest 0.5C step by the thermostat as most people don't care to get 0.1C level precision, while HA might get the actual value.

**Comment by rconte84** [comment]
@alexruffell IMHO this integration must show same values of Phisical Device/Nest app/Google Home and i think @Woodcat64 is saying this. For example, if my current home temp is show as 20.5° and target is 21° i don't want that HA show them like 20.4° and 20.8°

**Comment by alexruffell** [comment]
@rconte84 I have to disagree. If the device provides higher precision, then that is what should be pumped into HA. You can then create a template to round it up/down as you prefer. 

Using your example: current home temp shown as 20.5° (but it likely isn't exactly 20.5 because of the rounding) and target is 21°, then the best behavior is for HA to show the actual home temperature, let's say 20.4° and 21° as your target (the target should never be rounded - no need to anyway since you have to set a value in 0.5C steps - and should always match what you picked). If the sensors were very inaccurate, then rounding would be ok as you are not losing any information but the sensors I have (2 thermostats and 12 room sensors) are actually quite accurate when compared to other sensors I purchased and/or built (esphome).

(Riccardo, sono emigrato in USA nel 2001 ma sono di Roma! Ciao!)

**Comment by rconte84** [comment]
@alexruffell I don't know as well HA so it's possible i'm wrong... perhaps it's not a core issue but something related to the specific integration with Nest (the native assence of rounded values).
By the way i'll resolve with a template so my Google Nest data don't differ from HA.

PS: Forza Roma! Io sono toscano ;-)

**Comment by alexruffell** [comment]
Closing as I finally got round to reporting this under HA core.
